doc/book/cookbook/setting-locale-depending-routing-parameter.md

```diff
@@ -99,247 +99,56 @@ return [
-    $app->attachRouteResultObserver($container->get(LocalizationObserver::class));
-    return $app;
-});
-```
-
-and there are ways to accomplish it in Aura.Di as well — but they're all 
-different, making the approach non-portable.
-
-### Extend the Application factory
-
-Alternately, extend the Application factory:
-
-```php
-class MyApplicationFactory extends ApplicationFactory
-{
-    public function __invoke($container)
-    {
-        $app = parent::__invoke($container);
-        $app->attachRouteResultObserver($container->get(LocalizationObserver::class));
-        return $app;
-    }
-}
-```
-
-Then alter the line in `config/autoload/dependencies.global.php` that registers 
-the `Application` factory to point at your own factory.
-
-This approach will work across all container types, and is essentially a 
-portable way of doing delegator factories.
-
-### Use middleware
-
-Alternately, use `pre_routing` middleware to accomplish the task; the middleware 
-will get both the observer and application as dependencies, and simply register 
-the observer with the application:
-
-```php
-use Zend\Expressive\Router\RouteResultSubjectInterface;
-
-class LocalizationObserverMiddleware
-{
-    private $application;
-    private $observer;
-
-    public function __construct(LocalizationObserver $observer, RouteResultSubjectInterface $application)
-    {
-        $this->observer = $observer;
-        $this->application = $application;
-    }
-
-    public function __invoke($request, $response, callable $next)
-    {
-        $this->application->attachRouteResultObserver($this->observer);
+        $locale = $request->getAttribute('locale', 'de_DE');
+        Locale::setDefault($locale);
         return $next($request, $response);
     }
 }
 ```
 
-The factory would inject the observer and application instances; we leave this
-as an exercise to the reader.
-
-In your `config/autoload/middleware-pipeline.global.php`, you'd do the following:
+In your `config/autoload/middleware-pipeline.global.php`, you'd register the
+dependency, and inject the middleware into the pipeline following the routing
+middleware:
 
 ```php
 return [
     'dependencies' => [
-        'factories' => [
-            LocalizationObserverMiddleware::class => LocalizationObserverMiddlewareFactory::class,
+        'invokables' => [
+            LocalizationMiddleware::class => LocalizationMiddleware::class,
             /* ... */
         ],
         /* ... */
     ],
     'middleware_pipeline' => [
-        'pre_routing' => [
-            [ 'middleware' => LocalizationObserverMiddleware::class ],
-            /* ... */
-        ],
-        'post_routing' => [
-            /* ... */
+        /* ... */
+        [
+            'middleware' => [
+                Zend\Expressive\Container\ApplicationFactory::ROUTING_MIDDLEWARE,
+                Helper\UrlHelperMiddleware::class,
+                [ 'middleware' => LocalizationMiddleware::class ],
```

Comment:
actually I have a doubt. Is [] mean optional or we can write middleware aarays inside middleware ?